File: Python/stack/stack.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class Stack:
    def __init__(self):
        self.top = None
        self.height = 0

    # push a value to the top of the stack
    def push(self, value):
        new_node = Node(value)
        new_node.next = self.top
        self.top = new_node
        self.height += 1
    
    # pop the top value from the stack
    def pop(self):
        if self.height == 0:
            return None
        value = self.top.value
        self.top = self.top.next
        self.height -= 1 
        return value
    
    # peek the top value from the stack
    def peek(self):
        if self.height ==0:
            return None
        return self.top.value

    # print the stack
    def print_stack(self):
        temp = self.top
        print("Stack:")
        while temp is not None:
            print(temp.value)
            temp = temp.next
        print("End of Stack")

File: Golang/stack/stack.go
package stack

// Stack is an interface that defines the methods for a stack
type Stack interface {
	Push(value int)
	Pop() int
	Peek() int
	Height() int
	Print()
}

type node struct {
	value int
	next  *node
}

type stack struct {
	top    *node
	height int
}

// NewStack creates a new stack with the given values
// using a linked list stack implementation is more efficient if we don't know the size of the stack
// and we need to resize the underlying array and copy the values which is expensive
func NewStack(values []int) Stack {
	stack := &stack{}
	for _, value := range values {
		stack.Push(value)
	}
	return stack
}

type arrayStack struct {
	values []int
	length int
}

// NewArrayStack creates a new array stack with the given values and size
// using an array stack implementation is more efficient if we know the size of the stack
// and we don't need to resize the underlying array
func NewArrayStack(values []int, size int) Stack {
	stack := &arrayStack{
		values: make([]int, 0, size),
		length: 0,
	}
	for _, value := range values {
		stack.Push(value)
	}
	return stack
}
